File: src/model/rules/gameRule/GameRule.java
package model.rules.gameRule;

import model.enums.GameType;
import model.exceptions.GameException;
import model.exceptions.InvalidMoveException;
import model.pieces.Piece;
import model.structs.GameStatistics;
import model.structs.Move;

public interface GameRule {

    GameType getGameType();

    Piece[][] initializeGrid(int height, int width);

    void initializeExtraInfo(GameStatistics statistics);

    boolean placePieceValidationCheck(Move move, GameStatistics statistics) throws GameException;

    void nextPlayer(GameStatistics statistics);

    boolean placePiece(Move move, GameStatistics statistics) throws GameException;

    boolean gameOverCheck(GameStatistics statistics);

    int getWhiteScore(GameStatistics statistics);

    int getBlackScore(GameStatistics statistics);
}

File: src/model/rules/gameRule/GameRule.kt
package model.rules.gameRule

import model.enums.GameType
import model.exceptions.GameException
import model.pieces.Piece
import model.structs.GameStatistics
import model.structs.Move

interface GameRule {
    val gameType: GameType

    fun initializeGrid(height: Int, width: Int): Array<Array<Piece>>

    fun initializeExtraInfo(statistics: GameStatistics)

    @Throws(GameException::class)
    fun placePieceValidationCheck(move: Move, statistics: GameStatistics): Boolean

    fun nextPlayer(statistics: GameStatistics)

    @Throws(GameException::class)
    fun placePiece(move: Move, statistics: GameStatistics): Boolean

    fun gameOverCheck(statistics: GameStatistics): Boolean

    fun getWhiteScore(statistics: GameStatistics): Int

    fun getBlackScore(statistics: GameStatistics): Int
}
